clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerInterceptors.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+
+import org.apache.kafka.clients.consumer.ConsumerInterceptor;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+import java.io.Closeable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A container that holds the list {@link org.apache.kafka.clients.consumer.ConsumerInterceptor}
+ * and wraps calls to the chain of custom interceptors.
+ */
+public class ConsumerInterceptors<K, V> implements Closeable {
+    private final List<ConsumerInterceptor<K, V>> interceptors;
+
+    public ConsumerInterceptors(List<ConsumerInterceptor<K, V>> interceptors) {
+        this.interceptors = interceptors;
+    }
+
+    /**
+     * This is called when the records are about to be returned to the user.
+     * <p>
+     * This method calls {@link ConsumerInterceptor#onConsume(ConsumerRecords)} for each
+     * interceptor. Records returned from each interceptor get passed to onConsume() of the next interceptor
+     * in the chain of interceptors.
+     * <p>
+     * This method does not throw exceptions. If any of the interceptors in the chain throws an exception,
+     * it gets caught and ignored, and next interceptor in the chain is called with 'records' returned by the
+     * previous successful interceptor onConsume call.
+     *
+     * @param records records to be consumed by the client.
+     * @return records that are either modified by interceptors or same as records passed to this method.
+     */
+    public ConsumerRecords<K, V> onConsume(ConsumerRecords<K, V> records) {
+        ConsumerRecords<K, V> interceptRecords = records;
+        for (ConsumerInterceptor<K, V> interceptor: this.interceptors) {
+            try {
+                interceptRecords = interceptor.onConsume(interceptRecords);
```

Comment:
We don't have any real use-cases for pipelining interceptors that mutate data. My reasoning is that we want to call every interceptor, even if some of them throws exceptions. One use-case for that is monitoring. Since we agreed (at least for the first implementation of interceptors) to not propagate interceptor exceptions, interceptors do not "fail" producing/consuming records; so we want to make sure we monitor all records, independent of other interceptors errors.
So I was considering two options:
1. Allow interceptors to continue to mutate records after some interceptors fail, as implemented now.
2. Same as option 1, but return original records from onConsume (same for ProducerInterceptor.onSend).
3. Rejected: Same as option 1, but if one of the interceptors throws an exception, pass original records to the next interceptor. This option is weird, because if say 3rd interceptor fails, the second will work on mutated records from the first interceptor, and the 4th will work on the original records.

I think realistically, our first version of interceptor implementation already limits the support of interceptors that mutate records, because we do not allow to propagate exceptions. Once we get more experience with interceptors, we will probably change that (or not). So for now, I think as long as we clearly document the behavior, I think we are fine. 

I chose option 1, because I think it is least weird (but this is very subjective, and I am ok with another option). The reason is that with option 2, if the second interceptor is read-only and it throws an exception, it will revert the mutation of the first interceptor. So, I propose to maybe improve javadoc with the caution that a support of pipeline of interceptors that mutate data is limited and has side-effects.